aten/src/ATen/native/sparse/cuda/SparseMatMul.cu

```diff
@@ -0,0 +1,884 @@
+
+template <class scalar_t>
+csrOutput cuSparse_matrix_multiply(
+    const csrMatrixRef<scalar_t>& A,
+    const csrMatrixRef<scalar_t>& B,
+    Tensor& output_values,
+    IntTensor& output_indices) {
+  
+  cudaDataType cuda_data_type = getTensorCudaDataType(output_values);
+
+  const int A_num_rows = A.size(0);
+  const int A_num_cols = A.size(1);
+  const int A_num_nnz = A.nnz_;
+
+  const int B_num_rows = B.size(0);
+  const int B_num_cols = B.size(1);
+  const int B_num_nnz = B.nnz_;
+
+  int* dA_csrOffsets = A.csr_pointers_;
+  int* dA_columns = A.csr_indices_;
+  scalar_t* dA_values = A.csr_values_;
+
+  int* dB_csrOffsets = B.csr_pointers_;
+  int* dB_columns = B.csr_indices_;
+  scalar_t* dB_values = B.csr_values_;
+
+  csrOutput out({A.size(0), B.size(1)});
+
+  out.csr_pointers_ = at::empty({out.size(0) + 1}, output_indices.options().dtype(kInt));
+
+  int* dC_csrOffsets = out.csr_pointers_.data_ptr<int>();
+  int* dC_columns = nullptr;
+  scalar_t* dC_values = nullptr; 
+
+  scalar_t alpha = 1.0f;
+  scalar_t beta = 0.0f;
+  cusparseOperation_t opA = CUSPARSE_OPERATION_NON_TRANSPOSE;
+  cusparseOperation_t opB = CUSPARSE_OPERATION_NON_TRANSPOSE;
+  cudaDataType computeType = cuda_data_type;
+
+  //--------------------------------------------------------------------------
+  // CUSPARSE APIs
+  cusparseHandle_t handle = NULL;
+  cusparseSpMatDescr_t matA, matB, matC;
+  void *dBuffer1 = NULL, *dBuffer2 = NULL;
+  size_t bufferSize1 = 0, bufferSize2 = 0;
+  TORCH_CUDASPARSE_CHECK(cusparseCreate(&handle));
+
+  // Create sparse matrix A in CSR format
+  TORCH_CUDASPARSE_CHECK(cusparseCreateCsr(
+      &matA,
+      A_num_rows,
+      A_num_cols,
+      A_num_nnz,
+      dA_csrOffsets,
+      dA_columns,
+      dA_values,
+      CUSPARSE_INDEX_32I,
+      CUSPARSE_INDEX_32I,
+      CUSPARSE_INDEX_BASE_ZERO,
+      cuda_data_type));
+  TORCH_CUDASPARSE_CHECK(cusparseCreateCsr(
+      &matB,
+      B_num_rows,
+      B_num_cols,
+      B_num_nnz,
+      dB_csrOffsets,
+      dB_columns,
+      dB_values,
+      CUSPARSE_INDEX_32I,
+      CUSPARSE_INDEX_32I,
+      CUSPARSE_INDEX_BASE_ZERO,
+      cuda_data_type));
+  TORCH_CUDASPARSE_CHECK(cusparseCreateCsr(
+      &matC,
+      A_num_rows,
+      B_num_cols,
+      0,
+      NULL,
+      NULL,
+      NULL,
+      CUSPARSE_INDEX_32I,
+      CUSPARSE_INDEX_32I,
+      CUSPARSE_INDEX_BASE_ZERO,
+      cuda_data_type));
+  //--------------------------------------------------------------------------
+  // SpGEMM Computation
+  cusparseSpGEMMDescr_t spgemmDesc;
+  TORCH_CUDASPARSE_CHECK(cusparseSpGEMM_createDescr(&spgemmDesc));
+
+  // ask bufferSize1 bytes for external memory
+  TORCH_CUDASPARSE_CHECK(cusparseSpGEMM_workEstimation(
+      handle,
+      opA,
+      opB,
+      &alpha,
+      matA,
+      matB,
+      &beta,
+      matC,
+      computeType,
+      CUSPARSE_SPGEMM_DEFAULT,
+      spgemmDesc,
+      &bufferSize1,
+      NULL));
+  
+  cudaMalloc((void**)&dBuffer1, bufferSize1);
```

Comment:
Instead of directly calling `cudaMalloc` to allocate the temporary buffers, you should use our caching allocator. Some examples of how to do this correctly at https://github.com/pytorch/pytorch/pull/46605/files

Reply:
(cudaMalloc is quite slow and the caching allocator helps us amortize calls multiple times)

Reply:
Another benefit to using the caching allocator api is you get RAII handling of deallocation, so it is impossible to forget. The code here will leak cuda memory if there is an error before you get to cudaFree.